### PR TITLE
feat(router): add mertics to apple pay flow

### PR DESCRIPTION
**Pull request by ShankarSinghC**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Currently there are two flows for apple pay (`manual` and `simplified`). This pr adds metrics by which we can monitor payments create, succeeded and failed in the particular flow.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/payments.rs [hunk 1]
```diff
@@ -1279,6 +1282,14 @@ where
                 &Some(merchant_connector_account.clone()),
             )?;
 
+            let apple_pay_flow =
+                decide_apple_pay_flow(payment_method_type, is_apple_pay_predecrypt);
```
Comment (vspecky) [inline crates/router/src/core/payments.rs:1286]: Return an `Option<ApplePayFlow>` from `is_applepay_predecrypt` function and use it in subsequent code. No need for another function.

**Review by vspecky — CHANGES_REQUESTED** [review]

**Review by vspecky — DISMISSED** [review]

File: crates/router/src/core/payments.rs [hunk 2]
```diff
@@ -1107,44 +1110,47 @@ fn is_payment_method_tokenization_enabled_for_connector(
...
-                        api_models::payments::ApplePayCombinedMetadata::Simplified { .. } => true,
-                        api_models::payments::ApplePayCombinedMetadata::Manual { .. } => false,
-                    },
-                    api_models::payments::ApplepaySessionTokenMetadata::ApplePay(_) => false,
-                })
+) -> RouterResult<Option<enums::ApplePayFlow>> {
+    let apple_pay_predecrypt = merchant_connector_account.clone().and_then(|mca| {
+        let metadata = mca.get_metadata();
+        metadata.and_then(|apple_pay_metadata| {
+            let parsed_metadata: Result<api_models::payments::ApplepaySessionTokenData, _> =
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/payments.rs:1134]: please log the error here

File: crates/router/src/core/payments.rs [hunk 3]
```diff
@@ -1107,44 +1110,47 @@ fn is_payment_method_tokenization_enabled_for_connector(
...
-                parsed_metadata.ok().map(|metadata| match metadata.data {
-                    api_models::payments::ApplepaySessionTokenMetadata::ApplePayCombined(
-                        apple_pay_combined,
-                    ) => match apple_pay_combined {
-                        api_models::payments::ApplePayCombinedMetadata::Simplified { .. } => true,
-                        api_models::payments::ApplePayCombinedMetadata::Manual { .. } => false,
-                    },
-                    api_models::payments::ApplepaySessionTokenMetadata::ApplePay(_) => false,
-                })
+) -> RouterResult<Option<enums::ApplePayFlow>> {
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/payments.rs:1130]: here need not return `RouterResult`

File: crates/router/src/core/payments.rs [hunk 4]
```diff
@@ -1107,44 +1110,47 @@ fn is_payment_method_tokenization_enabled_for_connector(
         .unwrap_or(false))
 }
 
-fn is_apple_pay_predecrypt(
+fn decide_apple_pay_flow(
     payment_method_type: &Option<api_models::enums::PaymentMethodType>,
     merchant_connector_account: &Option<helpers::MerchantConnectorAccountType>,
-) -> RouterResult<bool> {
+) -> RouterResult<Option<enums::ApplePayFlow>> {
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/payments.rs:1116]: here need not return RouterResult

**Review by Sangamesh26 — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by vspecky — APPROVED** [review]

**Review by srujanchikke — APPROVED** [review]

**Pull request merged**